benchmark/scenarios/test/test_rate_measurer.py

```diff
@@ -0,0 +1,147 @@
+        """
+        Helper function that will send the desired number of request.
+
+        :param rate_measurer: The `RateMeasurer` we are testing.
+        :param num_requests: The number of request we want to send.
+        :param num_samples: The number of samples to collect.
+        """
+        for i in range(num_samples * num_requests):
+            rate_measurer.request_sent()
+
+    def receive_requests(self, rate_measurer, num_requests, num_samples):
+        """
+        Helper function that will receive the desired number of requests.
+
+        :param rate_measurer: The `RateMeasurer` we are testing.
+        :param num_requests: The number of request we want to receive.
+        :param num_samples: The number of samples to collect.
+        """
+        ignored = ""
+        for i in range(num_samples):
+            for i in range(num_requests):
+                rate_measurer.response_received(ignored)
+            rate_measurer.update_rate()
+
+    def failed_requests(self, rate_measurer, num_failures, num_samples):
+        """
+        Helper function that will result the desired number of response
+        failures.
+
+        :param rate_measurer: The `RateMeasurer` we are testing.
+        :param num_failures: The number of requests we want to fail.
+        :param num_samples: The number of samples to collect.
+        """
+        result = None
+        for i in range(num_samples):
+            for i in range(num_failures):
+                rate_measurer.request_failed(result)
+            rate_measurer.update_rate()
+
+    def increase_rate(self, rate_measurer, num_requests, num_samples):
+        """
+        Helper function that will increase the rate, sending the
+        desired number of request, and receiving the same
+        amount of them.
+
+        :param rate_measurer: The `RateMeasurer` we are testing.
+        :param num_requests: The number of request we want to make.
+        :param num_samples: The number of samples to collect.
+        """
+        self.send_requests(rate_measurer, num_requests, num_samples)
+        self.receive_requests(rate_measurer, num_requests, num_samples)
+
+    def test_rate_is_zero_when_no_samples(self):
+        """
+        When no samples have been collected, the rate should be 0.
+        """
+        r = RateMeasurer()
+        self.assertEqual(r.rate(), 0, "Expected initial rate to be zero")
+
+    def test_rate_is_lower_than_target_when_not_enough_samples(self):
+        """
+        When the number of samples collected is less than the sample
+        size, the rate should be lower than `target_rate`.
+        """
+        r = RateMeasurer()
+        target_rate = 5
+        num_samples = r.sample_size - 1
+
+        self.increase_rate(r, target_rate, num_samples)
+
+        self.assertTrue(r.rate() < target_rate)
+
+    def test_rate_is_correct_when_enough_samples(self):
+        """
+        A RateMeasurer should correctly report the rate when enough
```

Comment:
Write test method docstrings without the word "correctly" - describe what it is that is considered correct.